jax/_src/scipy/signal.py

```diff
@@ -146,3 +151,343 @@ def detrend(data, axis=-1, type='linear', bp=0, overwrite_data=None):
+                     nfft=None, detrend_type='constant', return_onesided=True,
+                     scaling='density', axis=-1, mode='psd', boundary=None,
+                     padded=False):
+  """LAX-backend implementation of `scipy.signal._spectral_helper`.
+
+  Unlike the original helper function, `y` can be None for explicitly
+  indicating auto-spectral (non cross-spectral) computation.  In addition to
+  this, `detrend` argument is renamed to `detrend_type` for avoiding internal
+  name overlap.
+  """
+  if mode not in ('psd', 'stft'):
+    raise ValueError(f"Unknown value for mode {mode}, "
+                     "must be one of: ('psd', 'stft')")
+
+  def make_pad(mode, **kwargs):
+    def pad(x, n, axis=-1):
+      pad_width = [(0, 0) for unused_n in range(x.ndim)]
+      pad_width[axis] = (n, n)
+      return jnp.pad(x, pad_width, mode, **kwargs)
+    return pad
+
+  boundary_funcs = {
+      'even': make_pad('reflect'),
+      'odd': odd_ext,
+      'constant': make_pad('edge'),
+      'zeros': make_pad('constant', constant_values=0.0),
+      None: lambda x, *args, **kwargs: x
+  }
+
+  if boundary not in boundary_funcs:
+    raise ValueError(
+        f"Unknown boundary option '{boundary}', "
+        f"must be one of: {list(boundary_funcs.keys())}")
+
+  # Check/ normalize inputs
+  nperseg = jax.core.concrete_or_error(int, nperseg,
+                                       "nperseg of windowed-FFT")
+  if noverlap is None:
+    noverlap = nperseg // 2
+  else:
+    noverlap = jax.core.concrete_or_error(int, noverlap,
+                                          "noverlap of windowed-FFT")
+  if nfft is None:
+    nfft = nperseg
+  else:
+    nfft = jax.core.concrete_or_error(int, nfft,
+                                      "nfft of windowed-FFT")
+  axis = jax.core.concrete_or_error(operator.index, axis,
+                                    "axis of windowed-FFT")
+
+  _check_arraylike("_spectral_helper", x)
+  if y is not None:
+    _check_arraylike("_spectral_helper", y)
+  x = jnp.asarray(x)
+  if y is not None:
+    y = jnp.asarray(y)
+    outdtype = jnp.result_type(x, y, np.complex64)
+  else:
+    outdtype = jnp.result_type(x, np.complex64)
+
+  if y is not None and mode != 'psd':
+    raise ValueError("two-argument mode is available only when mode=='psd'")
+
+  if y is not None:
+    # Check if we can broadcast the outer axes together
+    xouter = list(x.shape)
+    youter = list(y.shape)
+    xouter.pop(axis)
+    youter.pop(axis)
+    try:
+      outershape = jnp.broadcast_shapes(xouter, youter)
+    except ValueError as e:
+      raise ValueError('x and y cannot be broadcast together.') from e
+
+  # Special cases for size == 0
+  if y is None:
+    if x.size == 0:
+      return jnp.zeros(x.shape), jnp.zeros(x.shape), jnp.zeros(x.shape)
+  else:
+    if x.size == 0 or y.size == 0:
+      outshape = outershape + (min([x.shape[axis], y.shape[axis]]),)
+      emptyout = jnp.moveaxis(jnp.zeros(outshape), -1, axis)
```

Comment:
Could use `_src/util.tuple_insert` here.

Reply:
Same here.  Once `tuple_insert` supported negative indices we should use it here.

Reply:
done